[SourceKit][IDE] New SourceKit request to retrieve context type information

rintaro opened the pull request:
This is a new SourceKit request which receives a position in the source file, returns possible expected types and their members which can be referenced by "implicit member expression" syntax.

rdar://problem/44698136

benlangmuir commented:
The name "context info" is fairly general; do you think "type context info" would work?  Right now all the results are related to the semantic type of the context.  On the other hand, if you foresee us adding non-type information here it might be misleading.

benlangmuir commented on tools/swift-ide-test/swift-ide-test.cpp:
```diff
@@ -677,6 +681,66 @@ removeCodeCompletionTokens(llvm::MemoryBuffer *Input,
+  }
+  llvm::outs() << "found code completion token " << CodeCompletionToken
+               << " at offset " << Offset << "\n";
+  llvm::errs() << "found code completion token " << CodeCompletionToken
```
Huh, I wonder why we write this to both `outs()` and `errs()`.  Seems silly.  I guess it's fine to copy what code-completion does for now though.

rintaro replied:
Yeah, I just copied from code-completion code.
I believe `outs()` is for `FileCheck` and `errs()` is to see which test case failed from failure log.

rintaro commented:
One possible enhancement in my mind is adding "kind" of context. Like

```swift
foo(x: #TOKEN#)  // -> call argument
return #TOKEN#   // -> return value
if #TOKEN# {     // -> statement condition
```

But I'm not sure it's valuable. And I think "type context info" still works even after adding it.
Let's change it to `typecontextinfo`.

rintaro committed:
[CodeCompletion] Only analyze inner most context

Now that this passes all test cases we currently have. Stop analyzing
outer contexts because it may emit invalid context type.

rintaro committed:
[IDE] Rename CodeCompletionTypeContextAnalyzer to ExprContextAnalyzer

NFC

rintaro committed:
[IDE] Refactor ExprContextAnalyzer. NFC

We don't need to keep ExprFinder beyond analysis.

rintaro committed:
[IDE] Decouple ExprContextAnalyzer and ExprContextInfo

Just refactoring.
ExprContextInfo automatically construct analyzer and analyze().

rintaro committed:
[IDE] ExprContextAnalyzer::Analyze() doesn't need to return success

rintaro committed:
[IDE] Rename collectArgumentExpectation() to analyzeApplyExpr()

Removed unnecessary arguments.

rintaro committed:
[AST] Add shouldHideFromEditor() method

This replaces shouldHideDeclFromCompletionResults() in IDE.

rintaro committed:
[IDE] Move ExprContextAnalyser to its own file

NFC

rintaro committed:
[IDE] Move typeCheckContext() to ExprContextAnalysis.cpp

NFC

rintaro committed:
[IDE] Add TypeContextInfo callbacks to parser

TypeContextInfo is for new SourceKit request which receives source
location, and returns context type and members which can be referenced
by "implicit member expression" syntax.

Implement that as a code completion callbacks.

rintaro committed:
[SourceKit] New request to retrieve context type information

This is a new SourceKit request which receives a position in the source
file, returns possible expected types and their members which can be
referenced by "implicit member expression" syntax.

rintaro commented:
@swift-ci Please smoke test

rintaro commented:
@swift-ci Please smoke test

benlangmuir commented on lib/AST/Decl.cpp:
```diff
@@ -2550,6 +2550,34 @@ bool ValueDecl::isUsableFromInline() const {
 }
 
+/// Returns \c true if this declaration is *not* intended to be used directly
+/// by application developers despite of the visibility.
```
We don't usually copy doc comments to the definition since then we have two places to keep up to date separately.

benlangmuir commented on lib/IDE/CodeCompletion.cpp:
```diff
@@ -2116,7 +2049,7 @@ class CompletionLookup final : public swift::VisibleDeclConsumer {
     if (!VD->hasName() ||
         !VD->isAccessibleFrom(CurrDeclContext) ||
-        shouldHideDeclFromCompletionResults(VD))
+        VD->shouldHideFromEditor())
```
Is the function `shouldHideDeclFromCompletionResults`  removed? I can't find it in this patch.

benlangmuir commented on lib/IDE/TypeContextInfo.cpp:
```diff
@@ -0,0 +1,267 @@
+
+  // Ignore callbacks for suffix completions
+  // {
+  void completeDotExpr(Expr *E, SourceLoc DotLoc) override{};
```
Style: should have a space before `{}`.

benlangmuir commented on tools/SourceKit/lib/SwiftLang/SwiftTypeContextInfo.cpp:
```diff
@@ -0,0 +1,246 @@
+  }
+  CI.performSema();
+
+  return true;
```
We should copy over the `trace::TracedOperation` code from code-completion as well so that this action shows up in compilation traces such as the Index Log.
